base/opaque_closure.jl

```diff
@@ -110,3 +110,197 @@ function generate_opaque_closure(@nospecialize(sig), @nospecialize(rt_lb), @nosp
     return ccall(:jl_new_opaque_closure_from_code_info, Any, (Any, Any, Any, Any, Any, Cint, Any, Cint, Cint, Any, Cint, Cint),
         sig, rt_lb, rt_ub, mod, src, lineno, file, nargs, isva, env, do_compile, isinferred)
 end
+
+struct Slot{T} end
+struct Splat{T}
+    value::T
+end
+
+# Args is a Tuple{Vararg{Union{Slot{T},Some{T}}}} where Slot{T} represents
+# an uncurried argument slot, and Some{T} represents an argument to curry.
+@noinline @generated function Core.OpaqueClosure(Args::Tuple, ::Slot{RT}) where RT
+    AT = Any[]
+    call = Expr(:call)
+    extracted = Expr[]
+    closure_args = Expr(:tuple)
+    for (i, T) in enumerate(Args.parameters)
+        v = Symbol("arg", i)
+        is_splat = T <: Splat
+        if is_splat # TODO: check position
+            push!(call.args, :($v...))
+            T = T.parameters[1]
+        else
+            push!(call.args, v)
+        end
+        if T <: Some
+            push!(extracted, :($v = something(Args[$i])))
+        elseif T <: Slot
+            SlotT = T.parameters[1]
+            push!(AT, is_splat ? Vararg{SlotT} : SlotT)
+            push!(closure_args.args, call.args[end])
+        else @assert false end
+    end
+    AT = Tuple{AT...}
+    return Base.remove_linenums!(quote
+        $(extracted...)
+        $(Expr(:opaque_closure, AT, RT, RT, #= allow_partial =# false, :(($(closure_args))->@inline $(call))))
+    end)
+end
+
+"""
+    TypedCallable{AT,RT}
+
+TypedCallable provides a wrapper for callable objects, with the following benefits:
+    1. Enforced type-stability (for concrete AT/RT types)
+    2. Fast calling convention (frequently < 10 ns / call)
+    3. Normal Julia dispatch semantics (sees new Methods, etc.) + invoke_latest
+    4. Full pre-compilation support (including `--trim` compatibility)
+
+## Examples
+
+```julia
+const callbacks = @TypedCallable{(::Int,::Int)->Bool}[]
+
+register_callback!(callbacks, f::F) where {F<:Function} =
+    push!(callbacks, @TypedCallable f(::Int,::Int)::Bool)
+
+register_callback!(callbacks, (x,y)->(x == y))
+register_callback!(callbacks, (x,y)->(x != y))
+
+# Calling a random (or runtime-known) callback is fast!
+@btime callbacks[rand(1:2)](1,1)
+```
+
+# Extended help
+
+### As an invalidation barrier
+
+TypedCallable can also be used as an "invalidation barrier", since the caller of a
+TypedCallable is not affected by any invalidations of its callee(s). This doesn't
+completely cure the original invalidation, but it stops it from propagating all the
```

Comment:
Maybe useful to note that this will also block other information propagation? I am thinking constant-propagation, effects, escape-analysis etc...